### PR TITLE
[Merged by Bors] - feat(Pointwise): monotonicity of `pow`

### PR DESCRIPTION
**Pull request by YaelDillies**
... and `s ^ n = ∅ ↔ s = ∅ ∧ n ≠ 0`

From GrowthInGroups

---

- [x] depends on: #19249

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Review by b-mehta — COMMENTED** [review]
This PR has many changes, including some name changes without deprecations, some deprecations, some implicitness changes, new lemmas, and some proof modifications. Please split this one up into smaller ones, and use the PR description to describe the changes you are making in more detail than "other lemmas"

**Comment by YaelDillies** [comment]
What about now?

File: Mathlib/Algebra/Group/Pointwise/Finset/Basic.lean [hunk 1]
```diff
@@ -946,29 +949,70 @@ protected def monoid : Monoid (Finset α) :=
...
+-- `pow_left_mono`
+
 @[to_additive]
-theorem pow_mem_pow (ha : a ∈ s) : ∀ n : ℕ, a ^ n ∈ s ^ n
-  | 0 => by
-    simp only [pow_zero, mem_one]
-  | n + 1 => by
-    simp only [pow_succ]
-    exact mul_mem_mul (pow_mem_pow ha n) ha
+protected lemma pow_right_mono (hs : 1 ∈ s) : Monotone (s ^ ·) :=
```
Comment (b-mehta) [inline Mathlib/Algebra/Group/Pointwise/Finset/Basic.lean:956]:
```suggestion
protected lemma pow_right_monotone (hs : 1 ∈ s) : Monotone (s ^ ·) :=
```
(as is explicitly mentioned in the naming convention!)

File: Mathlib/Algebra/Order/Monoid/Unbundled/Pow.lean [hunk 2]
```diff
@@ -119,6 +119,20 @@ theorem Right.pow_lt_one_of_lt {n : ℕ} {x : M} (hn : 0 < n) (h : x < 1) : x ^
 
 @[deprecated (since := "2024-09-21")] alias Right.pow_neg := Right.nsmul_neg
 
+/-- This lemma is useful in non-cancellative monoids, like sets under pointwise operations. -/
+@[to_additive
+"This lemma is useful in non-cancellative monoids, like sets under pointwise operations."]
+lemma pow_le_pow_mul_of_sq_le_mul [MulLeftMono M] {a b : M} (hab : a ^ 2 ≤ b * a) :
+    ∀ {n}, 1 < n → a ^ n ≤ b ^ (n - 1) * a
```
Comment (b-mehta) [inline Mathlib/Algebra/Order/Monoid/Unbundled/Pow.lean:126]: This lemma also works for `n = 1`

File: Mathlib/Algebra/Group/Pointwise/Finset/Basic.lean [hunk 3]
```diff
@@ -946,29 +949,70 @@ protected def monoid : Monoid (Finset α) :=
...
 
 @[to_additive]
-theorem pow_subset_pow (hst : s ⊆ t) : ∀ n : ℕ, s ^ n ⊆ t ^ n
-  | 0 => by
-    simp [pow_zero]
-  | n + 1 => by
-    rw [pow_succ]
-    exact mul_subset_mul (pow_subset_pow hst n) hst
+lemma pow_subset_pow_mul_of_sq_subset_mul (hst : s ^ 2 ⊆ t * s) (hn : 1 < n) :
+    s ^ n ⊆ t ^ (n - 1) * s := pow_le_pow_mul_of_sq_le_mul hst hn
```
Comment (b-mehta) [inline Mathlib/Algebra/Group/Pointwise/Finset/Basic.lean:977]: This one also works for `n = 1`

File: Mathlib/Algebra/Group/Pointwise/Finset/Basic.lean [hunk 4]
```diff
@@ -946,29 +949,70 @@ protected def monoid : Monoid (Finset α) :=
...
+
+@[to_additive (attr := simp) nsmul_empty]
+theorem empty_pow (hn : n ≠ 0) : (∅ : Finset α) ^ n = ∅ := by
+  rw [← tsub_add_cancel_of_le (Nat.succ_le_of_lt <| Nat.pos_of_ne_zero hn), pow_succ', empty_mul]
```
Comment (b-mehta) [inline Mathlib/Algebra/Group/Pointwise/Finset/Basic.lean:981]:
```suggestion
theorem empty_pow (hn : n ≠ 0) : (∅ : Finset α) ^ n = ∅ :=
  match n with | n + 1 => by simp [pow_succ]
```
imo this is quite a bit cleaner. 
(aside, there's also the slightly funny proof
```
simpa using pow_subset_pow_mul_of_sq_subset_mul (α := α) (pow_two ∅).le hn
```
)

File: Mathlib/Algebra/Group/Pointwise/Set/Basic.lean [hunk 5]
```diff
@@ -1328,6 +1375,16 @@ lemma preimage_mul (hm : Injective m) {s t : Set β} (hs : s ⊆ range m) (ht :
 
 end Mul
 
+section Monoid
+variable [Monoid α] [Monoid β] [FunLike F α β] [MonoidHomClass F α β]
+
+@[to_additive]
+lemma image_pow (f : F) (s : Set α) : ∀ n, f '' (s ^ n) = (f '' s) ^ n
+  | 0 => by simp [singleton_one]
+  | n + 1 => by simp [image_mul, pow_succ, image_pow]
```
Comment (b-mehta) [inline Mathlib/Algebra/Group/Pointwise/Set/Basic.lean:1384]: Can we also have the version of this resembling
`[MulHomClass F α β] (f : F) (s : Set α) : ∀ n, n ≠ 0 → f '' (s ^ n) = (f '' s) ^ n`?

File: Mathlib/Algebra/Group/Pointwise/Set/Basic.lean [hunk 6]
```diff
@@ -1357,6 +1414,12 @@ lemma preimage_div (hm : Injective m) {s t : Set β} (hs : s ⊆ range m) (ht :
 
 end Group
 
+variable {ι : Type*} {α : ι → Type*} [∀ i, InvolutiveInv (α i)]
+
+@[to_additive (attr := simp)]
+lemma inv_pi (s : Set ι) (t : ∀ i, Set (α i)) : (s.pi t)⁻¹ = s.pi fun i ↦ (t i)⁻¹ := by
+  simp_rw [← image_inv]; exact piMap_image_pi (fun _ _ ↦ inv_surjective) _
```
Comment (b-mehta) [inline Mathlib/Algebra/Group/Pointwise/Set/Basic.lean:1421]: This one doesn't need the inv to be involutive:
```
@[to_additive (attr := simp)]
lemma inv_pi (s : Set ι) (t : ∀ i, Set (α i)) : (s.pi t)⁻¹ = s.pi fun i ↦ (t i)⁻¹ := by ext x; simp
```

**Review by b-mehta — COMMENTED** [review]
Let's also add finset versions of `image_pow`

File: Mathlib/Algebra/Group/Pointwise/Finset/Basic.lean (hunk 4 above)
Comment (YaelDillies) [reply Mathlib/Algebra/Group/Pointwise/Finset/Basic.lean:981]: Wow yes that's a funny proof

**Review by b-mehta — APPROVED** [review]
Thanks

bors merge

**Pull request closed without merging**